lib/std/channels.nim

```diff
@@ -0,0 +1,397 @@
+# ----------------------------------------------------------------------------------
+
+type
+  ChannelRaw = ptr ChannelObj
+  ChannelObj = object
+    lock: Lock
+    notFullCond, notEmptyCond: Cond
+    closed: Atomic[bool]
+    size: int
+    itemsize: int # up to itemsize bytes can be exchanged over this channel
+    head: int     # Items are taken from head and new items are inserted at tail
+    tail: int
+    buffer: ptr UncheckedArray[byte]
+    atomicCounter: Atomic[int]
+
+  ChannelCache = ptr ChannelCacheObj
+  ChannelCacheObj = object
+    next: ChannelCache
+    chanSize: int
+    chanN: int
+    numCached: int
+
+# ----------------------------------------------------------------------------------
+
+proc numItems(chan: ChannelRaw): int {.inline.} =
+  result = chan.tail - chan.head
+  if result < 0:
+    inc(result, 2 * chan.size)
+
+  assert result <= chan.size
+
+template isFull(chan: ChannelRaw): bool =
+  abs(chan.tail - chan.head) == chan.size
+
+template isEmpty(chan: ChannelRaw): bool =
+  chan.head == chan.tail
+
+# Unbuffered / synchronous channels
+# ----------------------------------------------------------------------------------
+
+template numItemsUnbuf(chan: ChannelRaw): int =
+  chan.head
+
+template isFullUnbuf(chan: ChannelRaw): bool =
+  chan.head == 1
+
+template isEmptyUnbuf(chan: ChannelRaw): bool =
+  chan.head == 0
+
+# ChannelRaw kinds
+# ----------------------------------------------------------------------------------
+
+proc isUnbuffered(chan: ChannelRaw): bool =
+  chan.size - 1 == 0
+
+# ChannelRaw status and properties
+# ----------------------------------------------------------------------------------
+
+proc isClosed(chan: ChannelRaw): bool {.inline.} = load(chan.closed, moRelaxed)
+
+proc peek(chan: ChannelRaw): int {.inline.} =
+  (if chan.isUnbuffered: numItemsUnbuf(chan) else: numItems(chan))
+
+
+# Channels memory ops
+# ----------------------------------------------------------------------------------
+
+proc allocChannel(size, n: int): ChannelRaw =
+  result = cast[ChannelRaw](c_malloc(csize_t sizeof(ChannelObj)))
+
+  # To buffer n items, we allocate for n
+  result.buffer = cast[ptr UncheckedArray[byte]](c_malloc(csize_t n*size))
+
+  initLock(result.lock)
+  initCond(result.notFullCond)
+  initCond(result.notEmptyCond)
+
+  result.closed.store(false, moRelaxed) # We don't need atomic here, how to?
+  result.size = n
+  result.itemsize = size
+  result.head = 0
+  result.tail = 0
+  result.atomicCounter.store(0, moRelaxed)
+
+
+proc freeChannel(chan: ChannelRaw) =
+  if chan.isNil:
+    return
+
+  if not chan.buffer.isNil:
+    c_free(chan.buffer)
+
+  deinitLock(chan.lock)
+  deinitCond(chan.notFullCond)
+  deinitCond(chan.notEmptyCond)
+
+  c_free(chan)
+
+# MPMC Channels (Multi-Producer Multi-Consumer)
+# ----------------------------------------------------------------------------------
+
+proc sendUnbufferedMpmc(chan: ChannelRaw, data: pointer, size: int, nonBlocking: bool): bool =
```

Comment:
Very minor thing of course, but given these are called internally with constants only, should `nonBlocking` be a `static bool`?

Reply:
Thanks, it can be a followup pr.